File: src/modules/organization/controllers/organization.controller.ts
import { Response } from 'express';
import { PaginationResponseDto } from '../../../shared/domain/dto/pagination-response.dto';
import { IRequest } from '../../../shared/domain/interfaces/request.interface';
import { UserRequest } from '../../../shared/types/user-request';
import { generateJsonError } from '../../../shared/utils/error';
import { OrganizationFindManyDto } from '../domain/dto/organization-find-many.dto';
import { OrganizationEntity } from '../domain/entities/organization.entity';
import { organizationFindManyService } from '../services/organization-find-many.service';
import { organizationFindOneService } from '../services/organization-find-one.service';
import { OrganizationCreateDto } from '../domain/dto/organization-create.dto';
import { organizationCreateService } from '../services/organization-create.service';
import { OrganizationUpdateDto } from '../domain/dto/organization-update.dto';
import { organizationUpdateService } from '../services/organization-update.service';
import { OrganizationUpdateStatusDto } from '../domain/dto/organization-update-status.dto';
import { organizationUpdateStatusService } from '../services/organization-update-status.service';
import { organizationDeleteService } from '../services/organization-delete.service';
import { organizationUpdateAvatarService } from '../services/organization-update-avatar.service';
import { organizationDeleteAvatarService } from '../services/organization-delete-avatar.service';

class OrganizationController {
  async findAll(request: IRequest, response: Response) {
    try {
      const params: OrganizationFindManyDto = new OrganizationFindManyDto(
        request.query,
      );
      const userRequest: UserRequest = request.user;
      const organizations: PaginationResponseDto<OrganizationEntity> =
        await organizationFindManyService.execute(params, userRequest);
      response.status(200).json(organizations);
    } catch (error) {
      generateJsonError(error, response);
    }
  }

  async findOne(request: IRequest, response: Response) {
    try {
      const id: number = Number(request.params.id);
      const userRequest: UserRequest = request.user;
      const organization = await organizationFindOneService.execute(
        id,
        userRequest,
      );
      response.status(200).json(organization);
    } catch (error) {
      generateJsonError(error, response);
    }
  }

  async create(request: IRequest, response: Response) {
    try {
      const dto: OrganizationCreateDto = request.body;
      const organization: OrganizationEntity =
        await organizationCreateService.execute(dto);
      response.status(201).json(organization);
    } catch (error) {
      generateJsonError(error, response);
    }
  }

  async updateAvatar(request: IRequest, response: Response) {
    try {
      const id: number = Number(request.params.id);
      const file: Express.Multer.File = request.file as Express.Multer.File;
      const userRequest: UserRequest = request.user;
      const organization: OrganizationEntity =
        await organizationUpdateAvatarService.execute(id, file, userRequest);
      response.status(200).json(organization);
    } catch (error) {
      generateJsonError(error, response);
    }
  }

  async deleteAvatar(request: IRequest, response: Response) {
    try {
      const id: number = Number(request.params.id);
      const userRequest: UserRequest = request.user;
      const organization: OrganizationEntity =
        await organizationDeleteAvatarService.execute(id, userRequest);
      response.status(200).json(organization);
    } catch (error) {
      generateJsonError(error, response);
    }
  }

  async update(request: IRequest, response: Response) {
    try {
      const id: number = Number(request.params.id);
      const dto: OrganizationUpdateDto = request.body;
      const userRequest: UserRequest = request.user;
      const organization: OrganizationEntity =
        await organizationUpdateService.execute(id, dto, userRequest);
      response.status(200).json(organization);
    } catch (error) {
      generateJsonError(error, response);
    }
  }

  async updateStatus(request: IRequest, response: Response) {
    try {
      const id: number = Number(request.params.id);
      const dto: OrganizationUpdateStatusDto = request.body;
      const userRequest: UserRequest = request.user;
      const organization: OrganizationEntity =
        await organizationUpdateStatusService.execute(id, dto, userRequest);
      response.status(200).json(organization);
    } catch (error) {
      generateJsonError(error, response);
    }
  }

  async delete(request: IRequest, response: Response) {
    try {
      const id: number = Number(request.params.id);
      const userRequest: UserRequest = request.user;
      const organization: OrganizationEntity =
        await organizationDeleteService.execute(id, userRequest);
      response.status(200).json(organization);
    } catch (error) {
      generateJsonError(error, response);
    }
  }
}

export const organizationController = new OrganizationController();

File: src/modules/organization/controllers/organization_controller.py
from ninja import Router, Query, File
from ninja.files import UploadedFile
from src.shared.decorators.require_abilities_decorator import require_abilities
from src.modules.role.domain.enums.ability_codes_enum import AbilityCodes
from src.modules.role.domain.enums.ability_actions_enum import AbilityActions
from src.shared.domain.dto.pagination_response_dto import PaginationResponseDTO
from src.shared.types.user_request import UserRequest
from src.modules.organization.domain.schemas.organization_schema import (
    OrganizationSchema,
)
from src.modules.organization.domain.dto.organization_find_many_dto import (
    OrganizationFindManyDTO,
)
from src.modules.organization.services.organization_find_many_service import (
    OrganizationFindManyService,
)
from src.modules.organization.services.organization_find_one_service import (
    OrganizationFindOneService,
)
from src.modules.organization.domain.dto.organization_create_dto import (
    OrganizationCreateDTO,
)
from src.modules.organization.services.organization_create_service import (
    OrganizationCreateService,
)
from src.modules.organization.domain.dto.organization_update_dto import (
    OrganizationUpdateDTO,
)
from src.modules.organization.services.organization_update_service import (
    OrganizationUpdateService,
)
from src.modules.organization.domain.dto.organization_update_status_dto import (
    OrganizationUpdateStatusDTO,
)
from src.modules.organization.services.organization_update_status_service import (
    OrganizationUpdateStatusService,
)
from src.modules.organization.services.organization_delete_service import (
    OrganizationDeleteService,
)
from src.modules.organization.services.organization_update_avatar_service import (
    OrganizationUpdateAvatarService,
)

from src.modules.organization.services.organization_delete_avatar_service import (
    OrganizationDeleteAvatarService,
)

router = Router(tags=["Organizations"])


@router.get("", response=PaginationResponseDTO[OrganizationSchema])
@require_abilities(AbilityCodes.ORGANIZATIONS, AbilityActions.READ)
def find_all(request, dto: Query[OrganizationFindManyDTO]):
    user_request: UserRequest = request.user
    return OrganizationFindManyService.execute(dto.dict(), user_request)


@router.get("/{id}", response=OrganizationSchema)
@require_abilities(AbilityCodes.ORGANIZATIONS, AbilityActions.READ)
def find_one(request, id: int):
    user_request: UserRequest = request.user
    return OrganizationFindOneService.execute(id, user_request)


@router.post("", response=OrganizationSchema)
@require_abilities(AbilityCodes.ORGANIZATIONS, AbilityActions.CREATE)
def create(request, dto: OrganizationCreateDTO):
    return OrganizationCreateService.execute(dto)


@router.put("/{id}", response=OrganizationSchema)
@require_abilities(AbilityCodes.ORGANIZATIONS, AbilityActions.UPDATE)
def update(request, id: int, dto: OrganizationUpdateDTO):
    user_request: UserRequest = request.user
    return OrganizationUpdateService.execute(id, dto, user_request)


@router.patch("/{id}/status", response=OrganizationSchema)
@require_abilities(AbilityCodes.ORGANIZATIONS, AbilityActions.UPDATE)
def update_status(request, id: int, dto: OrganizationUpdateStatusDTO):
    user_request: UserRequest = request.user
    return OrganizationUpdateStatusService.execute(id, dto, user_request)


@router.post("/{id}/avatar", response=OrganizationSchema)
@require_abilities(AbilityCodes.ORGANIZATIONS, AbilityActions.UPDATE)
def update_avatar(request, id: int, file: UploadedFile = File(...)):
    user_request: UserRequest = request.user
    return OrganizationUpdateAvatarService.execute(id, file, user_request)


@router.delete("/{id}/avatar", response=OrganizationSchema)
@require_abilities(AbilityCodes.ORGANIZATIONS, AbilityActions.DELETE)
def delete_avatar(request, id: int):
    user_request: UserRequest = request.user
    return OrganizationDeleteAvatarService.execute(id, user_request)


@router.delete("/{id}", response=OrganizationSchema)
@require_abilities(AbilityCodes.ORGANIZATIONS, AbilityActions.DELETE)
def delete(request, id: int):
    user_request: UserRequest = request.user
    return OrganizationDeleteService.execute(id, user_request)
